### ISSUE TITLE
OpenAPI spec for the REST API

### ISSUE THREAD

**Issue by taras** [backstage/backstage#2566]
## Context

An organization that I'm working with is considering adopting Backstage as their developer portal. One of the hard(ish) requirements that they have is that all services that we consume provide an API specification. This requirement exists because the infrastructure team needs to automate communication with services. If a specification is available, then they can use code generators to generate API clients for these services. 

## Feature Suggestion

It would be very helpful if there was a way to ensure that services integrated into Backstage come automatically with specifications. .NET, Rails & [FastAPI](https://fastapi.tiangolo.com) all provide a mechanism that automatically generates an OpenAPI definition from route definition. I'm not aware of any tools that work like this in Node.js world.

## Possible Implementation

[express-openapi](https://github.com/wesleytodd/express-openapi) is a middleware that generates *some* specification from routes. It doesn't produce complete specifications - I don't even know if this is possible.

This request could also be satisfied by a GraphQL API that wraps services that Backstages consumes. The only consideration is making sure that the design of the APIs covers plugin services as much as possible.

**Comment by andrewthauer** [comment]
I ran across this project the other day that could be of some use when using a code first approach: https://github.com/bee-travels/openapi-comment-parser.

However, I agree that using annotations or a middleware would be more preferable.

**Comment by stefanalund** [comment]
Does it have to be automated from the beginning, or could we start with a manually edited file?

**Comment by taras** [comment]
It doesn't have to be automated. Having an OpenAPI spec for services included out of the box could be a good start.

**Comment by stefanalund** [comment]
@taras wanna take a stab at it?

**Comment by taras** [comment]
I wouldn't mind but we need a way to ensure that they say current. Any ideas?

**Comment by kaos** [comment]
I think it looks like express-openapi indeed produces a complete specification, although you will have to write the spec for each route manually. The upside to that vs a completely manual spec on the side is that you get the spec closer to the implementation. You may also get validation middlewares for your routes (both for request and response data) amongst others, if desired.. ;)

**Comment by andrewthauer** [comment]
I tried out https://github.com/bee-travels/openapi-comment-parser recently and it will generate the openapi spec through JSDoc comments. It's still manual, but the spec ends up being right right above the code. I haven't tried it, but I suspect you could use some of the existing openapi tools like validation in conjunction with this approach.

**Comment by kaos** [comment]
looked a bit more at express-openapi, and I must say that I was stumped to learn that you had to put your implementation along with the spec.. I don't fancy the idea that the spec drives/dictates the implementation, it should be the other way around (the spec reflects the registered routes)

**Comment by taras** [comment]
@kaos I agree - spec and implementation should be separate. 

We've been experimenting with this inside of our Backstage instance for integrating with our platform. We've been able to decouple the frontend from backend development by using OpenAPI spec in the middle. We generate a client with [autorest](https://github.com/Azure/autorest) and use [Prism](https://github.com/stoplightio/prism) to mock service while in development. We're going to put it through the grind internally and then make an RFC when we're happy with the workflow.

**Comment by kaos** [comment]
Thanks for the update, looking forward to it :)

**Comment by freben** [comment]
Just bumping this to say that there's definite interest in this still! :)

**Comment by OrkoHunter** [comment]
Hey @taras,

> We're going to put it through the grind internally and then make an RFC when we're happy with the workflow.

Just curious if there is anything you'd like to share on your experience so far on this setup that you mention!

**Comment by taras** [comment]
@OrkoHunter yep, the setup that we have right now is this,

1. each service that we're consuming in Backstage has an OpenAPI spec
2. we generate clients for these services with [autorest](https://github.com/Azure/autorest)
3. the generated code is wrapped in a `use*API` hook which uses the Backstage's ref API
4. we use the `use*API` hooks in Backstage
5. we use [Prism](https://stoplight.io/open-source/prism/) to create a mock server in development - this allow us to build UI before backend microservice is implemented

This is pretty spiky at the moment because Autorest is quite heavy. We could create a leaner generator that generates hooks automatically and eliminate glue. Prism is also not exactly what we want to because it allows mocking by matching requests but not a high fidelity simulation. There is an issue to make [Prism more realistic](https://github.com/stoplightio/prism/issues/1224).

I would like to see and contribute to is having OpenAPI spec for each service in Backstage. Then generate clients wether at built time or runtime from the OpenAPI spec.

**Comment by alexcurtin** [comment]
@taras I'm very interested in seeing how you generated the OpenAPI spec for the catalog API, do you mind sharing your implementation somewhere? Can you push up a branch showcasing what you've built so far? Our end goal is to be able to build a catalog-client from the OpenAPI spec for the `/api/catalog` endpoint. Thanks!

**Comment by taras** [comment]
👋 @alexcurtin we haven't integrated with the catalog so far. We only used the clients that we created to integrate services that already have OpenAPI specs into Backstage. I might have an opportunity to do it this week, but I'm trying to figure out how to validate the implementation.

**Comment by taras** [comment]
I'm looking at using [json-schema-to-openapi-schema](https://github.com/openapi-contrib/json-schema-to-openapi-schema) to convert at least a part of https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/schema/kinds/API.v1alpha1.schema.json to YAML. From there, I'm going to try to create an OpenAPI spect that  includes the routes and references the generated files.

**Comment by freben** [comment]
Nice. Yeah if we could reduce the duplication of defining the entity shape it'd be nice. But why API? I'd have expected Entity instead, since that's the actual return type. Unless you wanna get fancy and do oneOf each of the specific kinds, but the value of that is probably very limited especially since it won't account for custom kinds etc

**Comment by taras** [comment]
> But why API? I'd have expected Entity instead, since that's the actual return type.

You're right. I referenced the wrong file. It was the last file opened in my tab instead of the proper file :) https://github.com/backstage/backstage/blob/master/packages/catalog-model/src/schema/Entity.schema.json is the right file.

**Comment by taras** [comment]
@freben I started creating a `@backstage/openapi` package. I'm going to put all of the generated YAML files in there. Does that sound alright?

**Comment by freben** [comment]
Uh, I am not sure. @Rugvip @benjdlambert @jhaals any input on that?

Gut feeling says that the catalog API, for one, is a thing of its own and should go in the `@backstage/catalog-client` package with the other client functionality. And that an overall mega-api-definition for a backend, rather than individual plugins, might be of limited use.

**Comment by freben** [comment]
Regarding the linked ticket about refs, input is welcome. I have run into annoying situations where `ajv` and other consumers treat refs in (sometimes subtly) different ways. Some like paths, some like identifiers, ...

We could update the ref usage but it would have to fit at least ajv and openapi at the same time.

**Comment by taras** [comment]
> We could update the ref usage but it would have to fit at least ajv and openapi at the same time.

I need to customize dereferencing to support identifiers. `json-schema-to-openapi-schema` also needs to be changed to support draft7. The rest of the conversion mechanism is quite simple so I might end up rewritinging the whole library.

> And that an overall mega-api-definition for a backend, rather than individual plugins, might be of limited use.

Yeah, this doesn't seem great. I'm in doing this in part to see if I can use the OpenAPI spec to generate a frontend client. We also want to make sure that anytime that schema changes, we regenerate the OpenAPI spec.

**Comment by freben** [comment]
Could it be useful to hand craft the api schema first, and add an integration test that builds a js client from the schema and performs calls against a running backend? :) That way we could at least reject the PR based on forgetting to update definitions - kinda like the api report stuff, one might say.

But yeah if you are far along towards solving schema generation from code, then sure that's even more convenient if it's a stable method.

**Comment by taras** [comment]
> Could it be useful to hand craft the api schema first, and add an integration test that builds a js client from the schema and performs calls against a running backend?

That was my original plan, but I didn't know how to validate my hand crafted OpenAPI spec, that's why I went down the path of generating from schema file. I started researching how I could generate a client and I came across something that might be interesting. 

Have you seen [Optic](https://useoptic.com)? It monitors test traffic to identify endpoints. It can [generate OpenAPI spec](https://useoptic.com/docs/apiops/openapi) from information that's entered in their UI. It can on PRs to show any changes to those contracts.

I'm going to give it a spin over the weekend.

**Comment by acunniffe** [comment]
> Have you seen [Optic](https://useoptic.com)? It monitors test traffic to identify endpoints. It can [generate OpenAPI spec](https://useoptic.com/docs/apiops/openapi) from information that's entered in their UI. It can on PRs to show any changes to those contracts.

👋 Optic maintainer here -- this seems really useful and I'm happy to help out @taras the community with some docs or guidelines for using Optic to document each plugin.

**Comment by taras** [comment]
@acunniffe 👋 I just got optic setup to start the backend and work as a proxy. It did exactly what I need it to - immediately. 

<img width="1818" alt="image" src="https://user-images.githubusercontent.com/74687/120205800-5e1de600-c1f8-11eb-90f8-1b01a8a9b6f8.png">

**Comment by taras** [comment]
@freben I want to make sure that I understand what would be beneficial at this stage. I'm going to describe some user stories that we want to be able to support.

1. As a developer building a portal with Backstage, I want to see the documentation for REST APIs provided by Backstage core.
2. As a contributor to a Backstage project, I want to verify that I did not change the implementation of a backend plugin in a way that breaks a documented API contract.
3. As a developer building a portal with Backstage, I want to be able to easily integrate an existing OpenAPI documented service into Backstage with all of the necessary TypeScript/React Hooks glue.

Did I get these right? Are there other use cases that I'm missing? 

For #1, as long as we have an OpenAPI spec that is current and accurate, we can use it to generate documentation.
For #2, we need some automated tests that make calls from the browser to the backend to ensure that the client and backend stay in sync - the contract can be verified by observing these tests
For #3, this needs to be created - right?

**Comment by freben** [comment]
The main use case I had in mind was

- As an engineer at our company, I want to make an integration with the Backstage catalog from other non-Backstage services written in other languages, and would like to automatically generate clients for it

This is a HUGE use case for us. The catalog powers, or at least informs, so much more than just Backstage itself.

And as a result of that, and to get a good dogfooding going, I was considering exploring whether catalog-client indeed also could make sense to auto generate. As well as exploring generating express stubs for the server side from it - in a schema-first fashion.

But I don't know the viability of all that yet

**Comment by taras** [comment]
I don't know either, but I want to figure out how to make it happen.

@freben how attached are you to the JSON Schema implementation that we have today? If there was another way to perform that validation, would you be open to it? I just want to understand what can be changed.

**Comment by taras** [comment]
I've been keeping my eye on https://github.com/aws/jsii/ for a while. This could be used to generate clients in different languages. 

Ok, I'm thinking that the first step could be to write a few e2e tests that use the catalog-client to interact with the catalog backend. We can use these requests to verify payloads that are sent to the backend API. Then use a proxy to inject a verification mechanism between the client and backend. The proxy will verify that the requests and responses conform to the spec.

Once we have this, then we can figure out how to proceed. How does this sound?

**Comment by jvilimek** [comment]
Hi guys. We are currently at the early stage of the Backstage implementation. And any such OpenAPI definitions and/or swagger help page would be really helpfull for us to see, what kind of API backstage and its plugins provide! Any news regarding this?

**Comment by freben** [comment]
Sorry, no news. It's just about having the time to soldier through making it - and our priorities have been elsewhere while moving everything toward GA release.

We're more than happy to help out with specifics on discord or elsewhere, and in a pinch you sometimes can glean details by looking at the source for the exposed routers of the plugins.

**Comment by TejasQ** [comment]
So we can probably use something like https://dev.to/kabartolo/how-to-document-an-express-api-with-swagger-ui-and-jsdoc-50do to autogenerate a nice documentation site. Just a suggestion.

**Comment by benjdlambert** [comment]
I wonder if we could get super meta, and use the `API` Entity and automatically add Backstage components to the `backstage` namespace so you can see these swagger docs in Backstage.

**Comment by alexcurtin** [comment]
Any news on this front?

**Comment by PhakornKiong** [comment]
I would love to help out on this as I'm trying to familiarize with the codebase. However, would like to know more on the direction on how to document OpenAPI spec.

**Comment by taras** [comment]
@PhakornKiong I, unfortunately, haven't been able to work on this. If you'd like, we can talk in Discord and I can tell you what I know.

**Comment by alexcurtin** [comment]
https://github.com/backstage/backstage/issues/7162 might be of interest.

**Comment by jvilimek** [comment]
Since the Backstage is GA...is this something that can be improved now? O:-)

**Comment by Rugvip** [comment]
@jvilimek We're currently doing some work that might touch on this, but not entirely sure where we'll end up or if we'll cover it. I'd say we're looking for ways were it can be more than just a decoupled schema only living in git.

**Comment by holwerda** [comment]
Any update on this?

**Comment by freben** [comment]
No. It'll remain a bit down the prio queue for some time, from the maintainer's point of view at least. If somebody comes up with a killer plan and implementation of how to achieve this, of course we and others will be very grateful.

**Comment by taras** [comment]
As the person who wrote the original issue, I want to update my view on this request. 

I still think it's very valuable, but I haven't done anything to address this issue because I don't want to add yet another file that needs to be maintained in addition to [@backstage/catalog-client](https://github.com/backstage/backstage/tree/master/packages/catalog-client). The process of extending the `@backstage/catalog-client` is already quite extensive as you can see by #14354 and #13583. Adding a spec file to requirements would just make it even more difficult.

With that said, I believe we could have best of both worlds: an OpenAPI spec and an up to date client. We can accomplish this by generating the `@backstage/catalog-client` from an OpenAPI spec. There are a few different tools for doing this, but [autorest](https://github.com/Azure/autorest) is one of the most popular options. It's a Node.js based CLI tool that generates clients in many languages. 

We could create a template similar to [`@autorest/typescript`](https://github.com/Azure/autorest.typescript/tree/main/packages/autorest.typescript) that would have Backstage specific client setup. This would make generating clients for Backstage projects very easy.

Autorest is written in Node.js, so we could even include this template as part of our `@backstage/cli` package.

This is obviously a lot of work. That's part of the reason why I haven't attempted to do it. If there are people who're interested in making it easier to integrate Backstage projects with external APIs, then I'd be happy to work with you to figure out how we can make this happen.

**Comment by sennyeya** [comment]
Hey all,
Wanted to chime in, I came across the `tsoa` package (https://github.com/lukeautry/tsoa) and did an initial PoC, https://github.com/sennyeya/backstage/blob/fb92c797e209aaac0badf4a1165390249d98b754/plugins/catalog-backend/src/swagger.json. Feel free to checkout the branch and poke around, the example-app should be working.

I think this would be a good first step to auto generate the OpenAPI documentation for the catalog. It has express integrations, but doesn't require full buy in. Each plugin can adopt it on their own using the `RegisterRoutes` method. The compilation is a little weird as it requires decorators and a specific file glob match pattern (https://tsoa-community.github.io/docs/getting-started.html). It also uses a more nest.js/Spring declarative pattern that may not be what we're looking for.

Curious to hear thoughts?

**Comment by pjungermann** [comment]
@taras at `plugins/bitbucket-cloud-common` we already use `@openapitools/openapi-generator-cli` to generate code from an OpenAPI specification.

**Comment by sennyeya** [comment]
@pjungermann I'm not sure I'm following, for #1. How would we capture business logic like pagination and validation with just API specifications? Wouldn't we still need code to capture this? ie something like `openapi-comment-parser` or `express-openapi.` In which case, we would be maintaining both anyways.

Once we have an API specification, we can use `autorest` or `@openapitools/openapi-generator-cli` to generate the clients. The PoC was more around looking at @taras's user stories 1 and 2,
1. As a developer building a portal with Backstage, I want to see the documentation for REST APIs provided by Backstage core.
2. As a contributor to a Backstage project, I want to verify that I did not change the implementation of a backend plugin in a way that breaks a documented API contract.
 - `tsoa` may not be a good fit here, but it would give an initial level of verification.

Re: plugins understanding their base paths: Makes sense. the plugin routers might have enough context to pull their subroutes, but it's something I'd have to take another look at.

I definitely agree, `tsoa`'s use of IoC is not great, but I think it would be required to access Backstage service's or the database in `tsoa`'s framework. I'm not advocating full buy in, it's just the best solution I've seen that would auto generate and enforce a spec.

Apologies if I'm missing something, I'm late to this thread and pretty new to the API spec/generated client world.

**Comment by sennyeya** [comment]
@Rugvip Hmm, I guess I'm still playing catchup. Here's my understanding at the moment, please correct me if/where I'm wrong,

I took a look at the existing OpenAPI express packages, there seem to be a few camps, schema-first or code-first. The schema-first ones require a full schema with endpoints and operations to do validation, `https://www.npmjs.com/package/express-openapi-validator`. They generally rely on operationIds being defined in the schema to define implementation. Code-first packages, like `tsoa`, generate either code and spec or just spec from typescript code or comments. There are also hybrid options, like `express-openapi` which supports both static file and a NextJS-like file structure.  The only major difference I see (and I could be off base) is that the process to generate the OpenAPI spec output is manual vs automatic. A framework like `tsoa` adds additional questions around IoC, but there are other options for code-first.

In which case, it seems to me like schema-first would have a higher cost especially if additional plugins don't use AJV (which at quick glance is most of them) or aren't well versed in OpenAPI. Also while both code-first and schema-first would require ongoing validation in PR pipelines, I would think that code-first would have more certainty that what you write is what you get.

Also, wouldn't a code-first approach like `tsoa` be more similar to schema-first than using an API spec file?

**Comment by pjungermann** [comment]
> How would we capture business logic like pagination and validation with just API specifications? Wouldn't we still need code to capture this? ie something like openapi-comment-parser or express-openapi. In which case, we would be maintaining both anyways.

You wouldn't do both. You express your contract regarding pagination inside of your OpenAPI specification.

The specification != implementation. As mentioned by @Rugvip you generate code from the spec, usually interfaces/types/... which then are used at the implementation.

Let's say, you use a pagination with limit and offset parameters, then these would be part of your spec incl. what you consider valid values for these. Your implementation then uses the defined parameters to provide pagination.
You don't need anything like openapi-comment-parser or express-openapi here.

> The PoC was more around looking at [...]

Sure. Don't take this as critique about your PoC. We need to look at different approaches I guess. Overall, code-first is an option. It just wouldn't be my preferred option however not sure what is the right approach for this project.

> The schema-first ones require a full schema with endpoints and operations to do validation, https://www.npmjs.com/package/express-openapi-validator.

Yes, we want a full schema :) Some of its validation requirements might not be necessary without using this specific package though.

> They generally rely on operationIds being defined in the schema to define implementation.

This package seems to be very open which comes with a price. They know about the spec and they know the code and need to connect both. Whether the code was manually written or generated with standard or custom code generator templates and whatever generator was used -- they don't rely on anything. Having operationIds is not something crazy though. Code-to-schema generators have them, too, just auto-generated values which might look a little weird sometimes.

> Also, wouldn't a code-first approach like tsoa be more similar to schema-first than using an API spec file?

No, this is just code-first, not schema-first. Starting with the schema (file) is what makes it schema-first. The schema file is also code/language agnostic. And you write and align on the contract before you start to implement it. Overall, I also believe that the quality of the schema will improve with schema-first. How good tooling can close the gaps might depend on the language, too.

The general challenge I see still with schema-first here is the fragmentation through plugins participating at the API which are mounted to a certain path prefix. And it should work to get one spec for a backstage deployment incl. custom plugins, etc.

Maybe something like https://www.npmjs.com/package/openapi-merge / https://www.npmjs.com/package/openapi-merge-cli could help.
Maybe https://www.npmjs.com/package/express-openapi works, too. I didn't look too deep into it. It seems like it can combine separately organized parts, however on a first glance it seemed to entangle spec parts and code a bit too much.
Probably, there are more alternatives.

A first try of openapi-merge-cli seems promising. Needs more exploration and would be just one part of the general plumbing.

Again, not sure if this is the right direction. Certainly, it may be one direction.

I think the concern about internal plugins is valid. The setup should support them.

> [...] or aren't well versed in OpenAPI

Not sure if this makes a huge difference for all approaches. It may cause issues with code-first as well.

If we consider a spec per plugin and then combine them in some way, it _may_ even be possible to allow code-first at internal plugins by Backstage users / a mixed approach. If the tooling/setup is good enough, it may not be needed though.

Overall, it seems quite big as pointed out before. Maybe we need a few more of these small PoCs to come up with a good plan.

**Comment by pjungermann** [comment]
@sennyeya I just pushed my try of openapi-merge to share how this merging could work

https://github.com/backstage/backstage/commit/0b6cb2f8de1f612ab9ca2f1f2a77cc84012c6d81

As the paths are customizable at each Backstage setup, the openapi-merge config file(s) could be part of the create-app command (initial skeleton and upgrade-diff).

This example is still very basic, however shows how this could work, even including customizations and custom/internal plugins, etc.

As `$.paths` needs to contain the full path, it is otherwise a bit tricky to break this down by sub-paths / path prefixes which is what is used here.

**Comment by pjungermann** [comment]
related to generators:
https://npmtrends.com/@openapitools/openapi-generator-cli-vs-autorest-vs-nswag-vs-openapi-generator-vs-openapi-typescript-codegen-vs-openapi3-generator-vs-swagger-client-vs-swagger-codegen-vs-swagger-ts-codegen

**Comment by taras** [comment]
I've looked at a bunch of different generators and based on that, I would say that there are at least 3 different considerations:

1. Tooling ecosystem around the generator - including the CLI command and how it's configured
2. Quality of the generated code: type correctness and library API
3. The templating system

Usually, 2 & 3 are connected, because depending on the templating system, it might not able to generate correct types.

My memory is big fuzzy and all of these tools are similarly named, but if I remember correctly, `@openapitools` ecosystem was very Java(ish). It generated weird types without return types - this might have changed so we need to check. I've tried a bunch of others, but I'd need to go look through my notes.

I mentioned autorest because it's the most mature in all of these areas. 
1. They allow configuring generation from markdown files. So you can describe the configuration of the generators in an MD files as part of documentation. 
2. Their default library is very mature but it might be too much for what we need.
3. Their templating system is really solid and flexible, but its a bit more complicated to implement.

If we decided to use autorest, we'd need to create our own TypeScript template based on their base template.

**Comment by sennyeya** [comment]
@pjungermann Just finished the static analysis piece which might be the missing gap between `openapi-merge` and dynamic route names, https://github.com/backstage/backstage/compare/master...sennyeya:backstage:open-api-spec-merge#diff-3c3f29b0bd2c69e389f72465709024dcebf6ffdeb02098c35921de9d2a10a54cR91. The example output in the linked comment is the actual return value from the example backend. There is definitely a better/less error prone way to do this :P, but for now it works on the example app.

My thinking was we could use the base (non-CLI) version of the package to merge and then leave it up to plugins to produce correct specs themselves. I'll take a look at connecting the two today.

Re: generators, I think there's also two camps, client and server generators. Client generators would be something like `autorest` that create the fetch/request tie ins and types. Server generators would be something like `express-openapi` that use the schema to validate and fill out the requests.

**Comment by pjungermann** [comment]
@taras at `plugins/bitbucket-cloud-common/templates/` I was also adding some templates on my own. The use case there was a bit limited though.

I have zero experience with autorest. I guess which one to use needs a good look in itself.

> I think there's also two camps, client and server generators.

There are generators which support both sides, too. E.g., the one I've mentioned.

Depending on your project, you might just need one though. As the provider of the API, you would use the server part, however you may also offer a set of generated clients/SDKs in different languages. Or consumers generate them on their end.

**Comment by sennyeya** [comment]
Alright, quick update on my side. I generated the default output from the `@autorest/typescript` package inside of the `catalog-client` folder and cloned the default template into the repo as `packages/client-generator`. [link](https://github.com/backstage/backstage/compare/master...sennyeya:backstage:schema-based-generation#diff-d26771ef4e9947e22ff04acc6f45a4dff14cb71371b32e840cac8b18302c2d29R1).

Looking at the generated code, it seems heavily reliant on the azure tool system. For our internal clients it would make sense to template it to match the system we're already set up, but for other language clients, especially corporate ones, that may cause more fragmentation than anything. I'll have to look at `openapi-generator` more, but an initial look seems promising, they have a wider array of supported languages _and frameworks_. The Java buy in isn't great, but I think the handlebars templates would make development and updates a lot easier. 

Couple of thoughts on the compatibility of the existing client to be generated,
- There's some pretty in depth merging of filter values, I'm not we'd want to capture that in a client, especially if we allow generation that we don't control, ie not through the `packages/client-generator`, from the OpenAPI spec.
- `requestOptional`, where the return value being 404 isn't an error. This shouldn't be an issue if we are generating plugin specs at run time from config, but I'm not sure how we'd denote it otherwise. OpenAPI has no way to denote optional paths. This may be a case for code-first creation of the schema.
- `requestIgnored`, similar to above, how would we denote this at the spec level?

**Comment by taras** [comment]
@sennyeya another alternative to `@autorest/typescript` can be https://github.com/karlvr/openapi-generator-plus. It uses handlebars for templates and no Java required.

**Comment by pjungermann** [comment]
you don't need Java for `@openapitools/openapi-generator-cli` either; it is written in Java though.

> Built-in templates are written in Mustache [...]. Beginning with version 4.0.0, we support experimental Handlebars and user-defined template engines via plugins.

**Comment by sennyeya** [comment]
> you don't need Java for @openapitools/openapi-generator-cli either; it is written in Java though.

I think you do need Java, custom generators are still stored as jar files and the [the download process](https://github.com/OpenAPITools/openapi-generator-cli/blob/master/lib/index.js#L79) still downloads the main package as a JAR binary. And the run process is a NestJS app  (which I've never seen as a CLI tool, pretty cool) on top of [a java command.](https://github.com/OpenAPITools/openapi-generator-cli/blob/aa303d0c75ec71bf8e5d05cf8cc716afb8f74e0e/apps/generator-cli/src/app/services/generator.service.spec.ts#L125)

_________________________________________

Quick progress update, I looked at 3 sets of things today,
1. OpenAPI SDK generators beyond `autorest` and `@openapitools/openapi-generator-cli`.
2. OpenAPI express wrappers/stub generators like `express-openapi` and `openapi-backend`.
3. JSON schema generators/validators.

@taras  `openapi-generator-plus` is interesting and definitely the next best contender of the bunch. One thing that I noticed is that its output seems fairly more verbose than the other two, [example](https://github.com/backstage/backstage/compare/master...sennyeya:backstage:openapi-generator-plus-poc?expand=1#diff-2250660b700580b1ea2994af68939de2180f8c5e52456c815779e05509c53600R45), but I like the first class typescript support. It does have a little less client language support than either other tool, only Java/Javascript/Swift. If our primary goal with the SDK generation is to auto generate plugin clients, starting with `@backstage/catalog-client`, it puts us in a good spot from a tooling perspective. Once we have publishable API specs, we could also just host them in either the repo or the `backstage.io` site to be pulled and let users create their own SDKs from whichever language/library they choose. This would come with the caveat that not all tools support all OpenAPI features and users would need to be educated on what tools would work with which schemas. [example](https://github.com/OpenAPITools/openapi-generator/issues/7720).

Couple of thoughts I've had from looking around:
1. OpenAPI 3.1 is, in a perfect world, what we're looking for because of out of the box support for JSON schema components. Unfortunately, very few of the tools geared towards OpenAPI support it and their timelines are a ways out.
2. A server solution like `openapi-backend` that plays well with express would allow for plugins to incrementally adopt an OpenAPI spec and instantly have request validation. I'll have to look and see how incremental this process would be, if it would allow for incremental route swapping or would be at the plugin level.
3. The latest JSON schema is also pretty cool, but has very little support. https://github.com/hyperjump-io/json-schema-validator is one of the few libraries that actually implement the feature `$id` AND `$ref`. I have a couple of concerns about key overwriting with the existing merge system due to it being completely `$ref` based. Question related to that:
  - Is it possible to have a plugin running on a different level of `@backstage/catalog-client`?
Again, this functionality would drop with OpenAPI 3.1, and that won't be for a while so it's not something to concentrate on.

Thoughts for next directions:
1. We probably want to (atleast initially) peg to OpenAPI 3.0, OpenAPI 3.1 has been out for a while, but still has pretty minimal support across all of the tooling we would need. No generator SDK libraries support OpenAPI 3.1 yet, `openapi-generator` has a project to track it https://github.com/orgs/OpenAPITools/projects/4/views/1, but this is primarily for tracking the main processing Java project and not the [child handlebars templates](https://github.com/OpenAPITools/openapi-generator/issues/13073). From what I can tell, those are going to be a ways out. Additionally, we want to be careful on dependencies we choose as some of the APIDevTools organization projects are at risk of losing their primary maintainer. Packages like [swagger-parser](https://github.com/APIDevTools/swagger-parser) and [json-schema-ref-parser](https://github.com/APIDevTools/json-schema-ref-parser) are soon to probably be [unmaintained](https://github.com/APIDevTools/json-schema-ref-parser/issues/285). There are alternatives, but very few support the newest version of JSON schema. I found https://github.com/hyperjump-io/json-schema-validator which seems very promising, but I haven't been able to fully test it yet.
2. We probably want to use [openapi-backend](https://github.com/anttiviljami/openapi-backend) or [express-openapi](https://github.com/kogosoftwarellc/open-api/tree/master/packages/express-openapi#readme). `openapi-backend` supports 3.1, express-openapi doesn't. Both support validation and operationId based routing. `express-openapi` it looks like doesn't support schema validation, https://github.com/kogosoftwarellc/open-api/tree/master/packages/openapi-request-validator, just request validation.
3. For this initial work, we'll probably want to:
a. double publish entity specs to JSON schema and then to OpenAPI using [json-schema-to-openapi-schema](https://github.com/wework/json-schema-to-openapi-schema) until OpenAPI 3.1 support exists in relevant packages. Currently this would capped at JSON schema v5, due to lack of support from [json-schema-transform](https://github.com/cloudflare/json-schema-tools/tree/master/workspaces/json-schema-transform) or 
b. just use OpenAPI component schemas. We [lost out](https://swagger.io/docs/specification/data-models/keywords/) on some [features](https://phil.tech/2022/bundling-openapi-with-javascript/), I can push for and do some updates in relevant packages, but I'm out of my element for a lot of this.

**Comment by Rugvip** [comment]
Wanna chime in and highlight that we feel that the ability to have type definitions generated from the schema is very important, in particular for the server implementation. The DX around building implementations for the endpoints is very important and not having them be typed is both risky and a big gap in the DX.

**Comment by taras** [comment]
What do you folks think about getting together to talk through this in a Discord meeting?

**Comment by sennyeya** [comment]
👍, definitely would be interested.

**Comment by sennyeya** [comment]
I shared this on Discord but posting here as well. I set up https://doodle.com/meeting/organize/id/bD1zz5Kd to track availability for a meeting, hoping to meet next week.

**Comment by sennyeya** [comment]
@freben Can we reopen this? Still working on #15667.

**Comment by sennyeya** [comment]
@regicsolutions There is already an [OpenAPI 3.1 spec](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/openapi.yaml) of the catalog API. Is that what you're looking for? 

I'm also currently working on #15667 which is a solution for the problem of modelling schema-first, using an OpenAPI spec and getting Typescript type restrictions (not runtime) on query/path parameters, request and response bodies and express router paths.

**Comment by regicsolutions** [comment]
@sennyeya my use case is related to building a Golang software template where i am looking generate go code from an OpenAPI spec as described [here](https://habr.com/en/post/571660/)

**Comment by sennyeya** [comment]
@regicsolutions I think the a part of the goal of this work would be a vended set of clients, but we're not there yet. You're welcome to use the spec for generating clients as needed, but expect the spec to change as it's still missing some catalog endpoints and the rest of the plugins don't have a spec at all. Also, the current catalog spec removes the `/catalog` or whatever prefix you're using, so you will need to add that back with something like [`openapi-merge`](https://github.com/robertmassaioli/openapi-merge).

**Comment by regicsolutions** [comment]
> @regicsolutions I think the a part of the goal of this work would be a vended set of clients, but we're not there yet. You're welcome to use the spec for generating clients as needed, but expect the spec to change as it's still missing some catalog endpoints and the rest of the plugins don't have a spec at all. Also, the current catalog spec removes the `/catalog` or whatever prefix you're using, so you will need to add that back with something like [`openapi-merge`](https://github.com/robertmassaioli/openapi-merge).

Thanks! Would I create a action to run the openapi-merge cli command within a software template? Was thinking I would just hit a swagger mock server?

**Comment by regicsolutions** [comment]
> @regicsolutions If your goal is a client Go application for your Backstage instance, I would recommend vending that internally through a Go package and not relying on a template. If you're looking to create a Go application from just any OpenAPI 3 spec, this isn't the right issue. The [link](https://habr.com/en/post/571660/) you shared should work, but I would be careful to check for OpenAPI 3.1 vs OpenAPI 3.0 compatibility as many tools don't yet work with the new 3.1 spec format.

Thanks @sennyeya my goal is to be able to scaffold a repository/go application based on a OpenAPI 3 spec. The link requires running cli commands wasn't sure how to best handle that in a template.

**Comment by sennyeya** [comment]
@freben @taras  Wanted to follow up on this as I have some free time and #15667 was finally merged ❤️. 

Since we have a spec for the catalog now, I want to take a look at generating the catalog-client. I'm going to focus primarily on [`openapi-generator`](https://github.com/OpenAPITools/openapi-generator) with [`openapi-generator-plus`](https://github.com/karlvr/openapi-generator-plus) as a backup. We'll definitely want to deep dive into generated code and what not once I have example clients.

**Comment by sennyeya** [comment]
@bforbis Stay tuned for #17535, I ended up switching down to OpenAPI 3.0 for that work for the same reasons you're mentioning. If you want to handle that on your side more quickly, you can pull the changes that I made to the spec to translate it from 3.1 to 3.0.

**Comment by sennyeya** [comment]
Quick update for those following this, 
1. #17535 has been merged and is part of the `1.20.0-next.0` release. This adds a new catalog provider to show a merged OpenAPI spec for your backend. The set of plugins it pulls from is manually defined and very few plugins have specs so far (limited to just catalog, search and todo rn).
2. As part of #17535, plugins that use the `createValidatedOpenApiRouter` method from `@backstage/backend-openapi-utils` will now have a `/api/${pluginId}/openapi.json` route that serves the full OpenAPI spec for this plugin, ie all spec paths are correctly mounted to `/api/${pluginId}`, etc.
3. #18185 and #19955 have been merged and can be used as part of `1.20.0-next.0`. The first PR adds some basic linting of OpenAPI specs to the project using Stoplight and the second PR adds test -> spec validation (and optionally, automatic updating) using Optic.
4. #17470 is working and building finally! There were some hiccups while trying to find a good path forward, but I think we've overcome those. Will be trying to push on this now that the OpenAPI router work has mostly died down.

Looking forward, there are a couple of interesting areas I'd like to start exploring,
1. JSON schema ref deduplication. Answering, how can we have a consistent `Entity` and `Error` definition across plugins?
2. Notification of breaking changes when updating the spec. Optic has some pretty nice support for this, should be straightforward to add.
3.  Starting to remove a lot of types from `catalog-backend`, `search-backend`, and `todo-backend` to start and replacing them with their generated counterparts.
4. API fuzzing. Building off of #17875 and #19955. This one solves two problems,
    a. How can we continue to protect against bad input and ensure stability of our APIs? Mainly looking at 500 errors and bad requests.
    b. How can we keep our specs up to date without manually iterating through all of the request/responses, either in test cases or in the spec itself?

Additionally, if folks are interested in writing API specs for their own plugins, I've started some documentation at https://github.com/backstage/backstage/blob/master/docs/openapi/test-case-validation.md and will be adding more soon.
